### DISCUSSION TITLE
Using jinja `include` tag not necessary if templated

### DISCUSSION THREAD

**Discussion by yan-hic** [apache/airflow#19435] [Q&A] [answered]
https://github.com/apache/airflow/blob/d3d2a38059a09635ebd0dda83167d404b1860c2a/airflow/providers/google/cloud/operators/bigquery.py#L2207
The above indicates that the `query` nested field in `configuration` parameter is templated if ending with `.sql`, meaning it is loaded before passed to jinja.

However there is currently no logic to read the `sql` file - there is only for reading a `.json` at the `configuration`-level.

That explains why the `include` tag was used in following example - I was wondering why initially
https://github.com/apache/airflow/blob/d3d2a38059a09635ebd0dda83167d404b1860c2a/airflow/providers/google/cloud/example_dags/example_bigquery_queries.py#L118

->
What is the intent ? Missing logic (preferred as more intuitive) or use jinja's `include` ? If it's the latter, the operator code and oc should remove the sql templating.

**Comment by josh-fell** [comment, marked as answer]
The `template_fields_renderers` attribute doesn't mean that a field is templated but rather how a field (or a particular portion of a templated field in this case) should be rendered in the UI using a corresponding Lexer.

This example elevates a method of using Jinja in a different manner than what's typically seen in Airflow. Instead of building logic to handle the `query` field contents in the operator, the solution provides an implicit flexibility to use a SQL file or explicit SQL string. Granted being implicit of course means it's not obvious. Perhaps this is why the Jinja `include` statement was specifically used in the example DAG.  The [`BigQueryInsertJobOperator` documentation](https://airflow.apache.org/docs/apache-airflow-providers-google/1.0.0/operators/cloud/bigquery.html#execute-bigquery-jobs) actually mentions this use case directly:
> If you want to include some files in your configuration you can use include clause of Jinja template language...
The included file can also use Jinja templates which can be useful in case of .sql files.


The example could just have easily passed in `"SELECT * FROM {{ DATASET }}.{{ TABLE }}"` as the `query` arg, but you could imagine a much more complicated SQL would be difficult to understand if rendered like this if users didn't have access to the underlying code:
![image](https://user-images.githubusercontent.com/48934154/140595339-15d659a8-1d6e-4b96-8db9-81857aad6b3f.png)

But is instead rendered in a far more readable fashion like below so `configuration.query.query` was added to `template_fields_renderers`.
![image](https://user-images.githubusercontent.com/48934154/140595372-348b68f5-406c-4194-9cc6-fd9825cbc9f7.png)

**Comment by potiuk** [comment]
Just to add to great answer by @josh-fell : I think the confusion comes from "renderer" word ambiguity,

* Template fields are rendered by JINJA templating engine to a string form
* Those "rendered" strings are then again "rendered" to HTML by pygments to make it visually better. 

`template_field_renderer` is about the latter.
